### PR TITLE
hide impls if trait bound is proven from env

### LINKED ISSUE #76: using impl candidates shadowed by where-bound overflows by using impl with only super trait bounds
https://rust.godbolt.org/z/8fzsP4MbE

```rust
trait Super {
    type SuperAssoc;
}

trait Trait: Super<SuperAssoc = Self::TraitAssoc> {
    type TraitAssoc;
}

impl<T, U> Trait for T
where
    T: Super<SuperAssoc = U>,
{
    type TraitAssoc = U;
}

fn overflow<T: Trait>() {
    let x: <T as Trait>::TraitAssoc;
}
```
normalizing `<T as Trait>::TraitAssoc` in the elaborated environment `[T: Trait, T: Super, <T as Super>::SuperAssoc = <T as Trait>::TraitAssoc]` has a single impl candidate, which uses the environment to normalize `<T as Trait>::TraitAssoc` to itself :3

### PR DESCRIPTION
**Pull request by lcnr**
AVERT YOUR EYES @compiler-errors

fixes https://github.com/rust-lang/trait-system-refactor-initiative/issues/76 and https://github.com/rust-lang/trait-system-refactor-initiative/issues/12#issuecomment-1865234925

this is kinda ugly and I hate it, but I wasn't able to think of a cleaner approach for now. I am also unsure whether we have to refine this filtering later on, so by making the change pretty minimal it should be easier to improve going forward.

r? @BoxyUwU

### CODE REVIEW

File: tests/ui/traits/next-solver/env-shadows-impls/param-candidate-shadows-project.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+impl<T> Foo for T {
+    type Assoc = i32;
+}
+
+impl<T> Bar for T where T: Foo<Assoc = i32> {}
+
+fn require_bar<T: Bar>() {}
+
+fn foo<T: Foo>() {
+    // Unlike the classic solver, the new solver previous projected
```
Comment (lcnr) [inline tests/ui/traits/next-solver/env-shadows-impls/param-candidate-shadows-project.rs:18]:
```suggestion
    // Unlike the classic solver, the new solver previously projected
```

**Review by BoxyUwU — APPROVED** [review]

**Comment by BoxyUwU** [comment]
r=me once you fix typo and the link

**Comment by jackh726** [comment]
This is super weird

**Pull request merged**